lib/builtins/deploy-delegates/cfn-deployer/helper.js

```diff
@@ -1,212 +1,132 @@
-    const retryConfig = {
-        base: 2000,
-        factor: 1.1,
-        maxRetry: 50
-    };
-    const retryCall = (loopCallback) => {
-        cloudformationClient.describeStack(stackId, (stackErr, stackResponse) => {
-            if (stackErr) {
-                return loopCallback(stackErr);
-            }
-            const stackStatus = R.view(R.lensPath(['StackStatus']), stackResponse);
-            const statusReason = R.view(R.lensPath(['StackStatusReason']), stackResponse);
-            const reasonMsg = statusReason ? `Status reason: ${statusReason}.` : '';
-            reporter.updateStatus(`Current stack status: ${stackStatus}... ${reasonMsg}`);
-            loopCallback(null, stackResponse);
-        });
-    };
-    const shouldRetryCondition = retryResponse => !retryResponse || retryResponse.StackStatus.endsWith('IN_PROGRESS');
-    retryUtils.retry(retryConfig, retryCall, shouldRetryCondition, (err, res) => callback(err, err ? null : res));
-}
+    _stackExists(stackId) {
+        if (!stackId) return false;
+        return this.cloudformationClient.describeStacks({ StackName: stackId }).promise()
+            .then(data => !data.Stacks[0].StackStatus.startsWith('DELETE_'))
+            .catch(() => false);
+    }
 
-function _getEndpointOutput(resp, callback) {
-    const endpoint = resp.Outputs.filter(output => output.OutputKey === 'SkillEndpoint');
-    if (!endpoint[0] || !endpoint[0].OutputValue) {
-        return callback('No "SkillEndpoint" result returned in the stack outputs. '
-            + 'Make sure you have a "SkillEndpoint" output in your template which points to the endpoint URL.');
+    async deployStack(stackId, stackName, templatePath, parameters, capabilities) {
+        const templateBody = fs.readFileSync(templatePath, 'utf-8');
+        const stackExists = await this._stackExists(stackId);
+        if (stackExists) {
+            this.reporter.updateStatus(`Updating stack (${stackId})...`);
+        } else {
+            this.reporter.updateStatus(`No stack exists or stack has been deleted. Creating cloudformation stack "${stackName}"...`);
+        }
+        const params = {
+            StackName: stackExists ? stackId : stackName,
+            TemplateBody: templateBody,
+            Parameters: parameters,
+            Capabilities: capabilities,
+        };
+
+        if (stackExists) {
+            return this.cloudformationClient.updateStack(params).promise();
+        }
+        return this.cloudformationClient.createStack(params).promise();
     }
-    callback(null, endpoint[0].OutputValue);
-}
 
-function _getStackResourcesFailureReason(reporter, cloudformationClient, stackId, callback) {
-    const errorsList = [];
-    reporter.updateStatus('Checking the failure details for the stack resources...');
-    cloudformationClient.describeStackResources(stackId, (err, stackResources) => {
-        if (err) {
-            return callback(err);
+    async waitForStackDeploy(stackId) {
```

Comment:
[Apply to all] This func needs jsdoc comment for public methods. Especially the type of the return object. e.g for this one you are returning {stackInfo} stackInfo is better to be annotated.